ethers-contract/src/contract.rs

```diff
@@ -291,12 +336,13 @@ impl<M: Middleware> Contract<M> {
 
         let tx = tx.into();
 
-        Ok(ContractCall {
+        Ok(ContractCallInternal {
             tx,
-            client: Arc::clone(&self.client), // cheap clone behind the Arc
+            client: self.client.clone(),
```

Comment:
we could make `ContractCallInternal` borrow our B here, by introducing a lifetime to it

I am concerned about the practicality of the `B: Clone` bound, for borrow types _without_ clone. are there commonly used borrow types that are _NOT_ clone? 🤔

Reply:
It can be the case when `B: Borrow<M>` is equal to `M` and `M` is not Clone, since `Borrow` has a blanket implemention for any `T: ?Sized`. Otherwise it would mostly be the standard smart pointers. So I'd say if all our `Middleware` implement `Clone` this is fine

Reply:
unfortunately, not all of our middleware implement Clone, IIRC. e.g. I think `SignerMiddleware<LedgerSigner, _>` does not 🤔

Reply:
Ah yeah right, I guess this will need to be specified in a doc somewhere that it is recommended wrapping `M` with a smart pointer or using references to avoid cloning data or for cases where that it is actually not possible

Reply:
I'm going to make a couple toy examples for when `B` is `M` or `&M`. (`&M` is `Clone` however, so that should be fine)

Reply:
![Screenshot 2023-01-27 at 12 19 12 PM](https://user-images.githubusercontent.com/10149425/215152256-898d606d-ad3e-4d52-ba22-3625a45feab8.png)

Reply:
I've also confirmed that `NonceManagerMiddleware` is not Clone

Reply:
so the place where I user would run into this is:

- instantiating a contract with an _owned_ **non-clone** middleware

Reply:
slight adjustment to screenshot. 3rd example is wrong, as M is `&`. corrected version here:

![Screenshot 2023-01-27 at 12 27 58 PM](https://user-images.githubusercontent.com/10149425/215154007-cbcafdce-6390-480e-8bbb-c010eb1751cd.png)

Reply:
Is that a big problem? Can't we pass these Middlewares by reference, instead of owned? And otherwise, just Arc them?

Reply:
> Is that a big problem? Can't we pass these Middlewares by reference, instead of owned? And otherwise, just Arc them?

it's a not-immediately-obivous issue that users will run into and ask about. It compounds with the ambiguity caused by `auto_imp(&, Box, Arc)` to make this more-powerful API extremely difficult to use

if you call `ContractInstance::new(address, &provider)` you get a type ambiguity around `M`. is `M` `Provider` or `&Provider`. Both implement `Middleware` and both can be borrowed from `&Provider`. Same ambiguity for `Arc`

The follow up work (will copy this to original post) to get this usable by beginner rustaceans is something like:
```
1. list further abstraction targets
2. refactor to use borrow
3. provide backwards compatible type aliases like `type Contract = ContractInstance<Arc<M>, M>`
4. remove `auto_impl(&, Box, Arc)` on `Middleware`
5. write migration guide
6. deprecate backwards-compatibility aliases
7. remove aliases
```